"Set up web push notifications with Firebase Cloud Messaging" tutorial (#27)
talkjs/tutorials: how-to-set-up-web-push-notifications-with-firebase/how-to-set-up-web-push-notifications-with-firebase.md

# Set up web push notifications with Firebase Cloud Messaging

In this tutorial, we'll show you how to use [Firebase Cloud Messaging](https://firebase.google.com/docs/cloud-messaging) (FCM) with TalkJS to send web push notifications to your users' browsers when they receive a new message in their TalkJS chat:

> <!-- TODO: add 1-demo.jpg -->

Reviewer: Love the illustrations! 😀

</div>

## Contents

- [Prerequisites](#prerequisites)
- [Set up Firebase](#set-up-firebase)
  - [Step 1: Initialize Firebase](#step-1-initialize-firebase)
  - [Step 2: Create a new Firebase app](#step-2-create-a-new-firebase-app)
  - [Step 3: Generate your public key](#step-3-generate-your-public-key)
- [Create your web application](#create-your-web-application)
  - [Step 1: Create a TalkJS chatbox](#step-1-create-a-talkjs-chatbox)
  - [Step 2: Add the Firebase event listener](#step-2-add-the-firebase-event-listener)
  - [Step 3: Create the service worker](#step-3-create-the-service-worker)
  - [Step 4: Style the app](#step-4-style-the-app)
- [Test your app](#test-your-app)
  - [Step 1: Deploy your application to Firebase hosting](#step-1-deploy-your-application-to-firebase-hosting)
  - [Step 2: Send a message from another user](#step-2-send-a-message-from-another-user)
- [Conclusion](#conclusion)

## Prerequisites

To follow along with this tutorial, you'll need:

- A [TalkJS account](https://talkjs.com/dashboard/login)

> - A [Firebase account](https://console.firebase.google.com/) with a Firebase project set up. If you don't yet have a Firebase project, follow [their guides](https://firebase.google.com/docs/guides?_gl=1*qyyupu*_up*MQ..*_ga*MTQxMDg5NTA2Ni4xNzM4MjQ2OTI5*_ga_CW55HF8NVT*MTczODI0NjkyOS4xLjAuMTczODI0NjkyOS4wLjAuMA..) to get started.

Reviewer: It looks like the link might still contain some session-specific info? And a suggestion to make the link text a tad bit more descriptive:

```suggestion
- A [Firebase account](https://console.firebase.google.com/) with a Firebase project set up. If you don't yet have a Firebase project, follow [the Firebase fundamentals guides](https://firebase.google.com/docs/guides) to get started.
```

Author: Oh yeah, thanks for catching that!

- Firebase Cloud Messaging configured for your TalkJS account. For instructions on how to set this up in the TalkJS dashboard, see [Configure Firebase Cloud Messaging](https://talkjs.com/docs/Features/Notifications/Mobile_Push_Notifications/Configure_FCM/).
- The [Firebase CLI](https://firebase.google.com/docs/cli#install_the_firebase_cli)

## Set up Firebase

In this section, you'll initialize Firebase Hosting with the [Firebase CLI](https://firebase.google.com/docs/cli#install_the_firebase_cli), create a new Firebase app and generate a key to encrypt push notifications.

### Step 1: Initialize Firebase Hosting

Log in to Firebase with the Firebase CLI if you haven't already:

```sh
firebase login
```

Next, create a new directory for your code for this project and run the following command inside it:

```sh
firebase init hosting
```

Answer the questions when prompted:

> - Select to associate your new project directory with an existing Firebase project or create a new one. If you create a new one, or you haven't yet set up Firebase Cloud Messaging in the TalkJS dashboard, see [Configure Firebase Cloud Messaging](https://talkjs.com/docs/Features/Notifications/Mobile_Push_Notifications/Configure_FCM/) in our docs.
> - Select to use the default (`public`) as your public directory
> - Select whether to configure as a single-page app (in this example we've selected "Yes", but you can use either)
> - Choose whether to set up automatic builds and deploys with GitHub
> 
> Once initialization is complete, you should see some config files in your directory, along with a `public` directory with an `index.html` file. We'll update this file in a later step.
> 
> ### Step 2: Create a new Firebase app
> 
> Next, open the [Firebase console](http://console.firebase.google.com/) and select your Firebase project. Create a new app inside the Firebase project:
> 
> - Open the [Firebase console](http://console.firebase.google.com/) and select your project
> - From the **Project Overview** page, click the **+** icon to add a new web app

Reviewer: Tiny edit for consistency: Some of the list items had full stops at the end, some didn't. I think the rule would be to add a full stop if it's a complete running sentence, so have suggested that here 🙂 

```suggestion
- Select to associate your new project directory with an existing Firebase project or create a new one. If you create a new one, or you haven't yet set up Firebase Cloud Messaging in the TalkJS dashboard, see [Configure Firebase Cloud Messaging](https://talkjs.com/docs/Features/Notifications/Mobile_Push_Notifications/Configure_FCM/) in our docs.
- Select to use the default (`public`) as your public directory.
- Select whether to configure as a single-page app (in this example we've selected "Yes", but you can use either).
- Choose whether to set up automatic builds and deploys with GitHub.

Once initialization is complete, you should see some config files in your directory, along with a `public` directory with an `index.html` file. We'll update this file in a later step.

### Step 2: Create a new Firebase app

Next, open the [Firebase console](http://console.firebase.google.com/) and select your Firebase project. Create a new app inside the Firebase project:

- Open the [Firebase console](http://console.firebase.google.com/) and select your project.
- From the **Project Overview** page, click the **+** icon to add a new web app.
```

Author: Thanks, yes agree with that

        apiKey: "",
        authDomain: "",
        projectId: "",
        storageBucket: "",
        messagingSenderId: "",
        appId: "",
      });
      let features = [
        "auth",
        "database",
        "firestore",
        "functions",
        "messaging",
        "storage",
        "analytics",
        "remoteConfig",
        "performance",
      ].filter((feature) => typeof app[feature] === "function");
      loadEl.textContent = `Firebase SDK loaded with ${features.join(", ")}`;
    } catch (e) {
      console.error(e);
      loadEl.textContent = "Error loading the Firebase SDK, check the console.";
    }

    const messaging = firebase.messaging(firebase.app());

    document.getElementById("requestToken").addEventListener("click", () => {
      messaging
        .getToken({
          vapidKey: "<FIREBASE_PUBLIC_KEY>", // update this with your Firebase project public key
        })
        .then((token) => {
          window.talkSession.setPushRegistration({
            provider: "fcm",
            pushRegistrationId: token,
          });

          document.getElementById("token").textContent = token;
        });
    });
  });
</script>
```

> Replace the placeholder object in `firebase.initializeApp` with the `firebaseConfig` object that you saved in Step 2 of the previous section.
> 
> Also, replace `<FIREBASE_PUBLIC_KEY>` with the public key that you saved in Step 3 of the previous section.

Reviewer: By no means essential, just a possible convenience ☺️ 

```suggestion
Replace the placeholder object in `firebase.initializeApp` with the `firebaseConfig` object that you saved in [Step 2 of the previous section](#step-2-create-a-new-firebase-app).

Also, replace `<FIREBASE_PUBLIC_KEY>` with the public key that you saved in [Step 3 of the previous section](#step-3-generate-your-public-key).
```

This service worker handles incoming messages and adds logging and custom notification text. It's currently configured to only show notifications when the web page is in the background, to prevent users getting notified when they are already viewing the chat. If you want to instead show notifications when the web page is in the foreground, copy the code inside `onBackgroundMessage` into the `onMessage` handler.

### Step 4: Style the app

Finally, let's add some styling to the page. In the `head` tag of `index.html`, add the following script:

```html
<style media="screen">
  body {
    /* background: white; */
    color: rgba(0, 0, 0, 0.87);
    font-family: Roboto, Helvetica, Arial, sans-serif;
    margin: 0;
    padding: 0;
  }
  .load {
    color: #525252;
    text-align: center;
    font-size: 13px;
  }
  #requestToken {
    margin: 0px 0px 12px;
    padding: 8px 16px;
    border-radius: 4px;
    background: #007df9;
    color: white;
    border: none;
    cursor: pointer;
  }
  @media (max-width: 600px) {
    body {
      margin-top: 0;
      background: white;
      box-shadow: none;
    }
  }
</style>
```

> This styles the **Request token** button and the loading

Reviewer: Some word(s) missing here still for a complete sentence

Author: Oops, good catch. Fixed

## Test your app

Let's verify that your push notification setup is working correctly by redeploying the application and sending a message.

### Step 1: Deploy your application to Firebase hosting

Redeploy your site to Firebase Hosting:

```sh
firebase deploy --only hosting
```

This will deploy your website to a URL like `https://<APP_NAME>.web.app`.

You should now see a TalkJS chatbox and a **Request token** button:

<!-- TODO: add 3-chatbox-with-token-button.jpg -->

<figure class="kg-image-card">
  <img class="kg-image" src="<URL>" alt="<ALT>"/>
</figure>

### Step 2: Send a message from another user

We've set up our service worker to only show notifications when the web page is in the background, to save users from getting unnecessary notifications when they are already viewing the chat. So to test notifications, you will need to switch to another tab and then send a message from a different user:

- On the device where you want to receive push notifications, go to `https://<APP_NAME>.web.app` and click the **Request token** button. Select to allow push notifications if prompted.

> - Switch to a different website in another different tab, or to a different app.

Reviewer: Tiny style thing: 'another different tab' sounds a bit double? So could do either of these terms instead of both

Author: Thanks, yeah think that was an editing issue. Fixed

> - Add a new message to the TalkJS `sample_conversation` conversation from the other user, `sample_user_alice`. A simple way to do this is to go to the **Chat UI** tab of the TalkJS dashboard](https://talkjs.com/dashboard) and add a message from the **Preview** chatbox UI:

Reviewer: ```suggestion
- Add a new message to the TalkJS `sample_conversation` conversation from the other user, `sample_user_alice`. A simple way to do this is to go to the **Chat UI** tab of the [TalkJS dashboard](https://talkjs.com/dashboard) and add a message from the **Preview** chatbox UI:
```

# Set up web push notifications with Firebase Cloud Messaging

In this tutorial, we'll show you how to use [Firebase Cloud Messaging](https://firebase.google.com/docs/cloud-messaging) (FCM) with TalkJS to send web push notifications to your users' browsers when they receive a new message in their TalkJS chat:

<!-- TODO: add 1-demo.jpg -->

<figure class="kg-image-card">

>   <img class="kg-image" src="Demo of the feature. The user, Alice, selects to allow push notifications and then clicks the 'Request token' button. She then switches to another tab. The other participant in the conversation, Sebastian, then sends a message. This triggers a push notification for Alice." alt="<ALT>"/>

Author: Oops, thanks, that was meant to go in the <ALT> placeholder. Fixed.

  - [Step 4: Style the app](#step-4-style-the-app)
- [Test your app](#test-your-app)
  - [Step 1: Deploy your application to Firebase hosting](#step-1-deploy-your-application-to-firebase-hosting)
  - [Step 2: Send a message from another user](#step-2-send-a-message-from-another-user)
- [Conclusion](#conclusion)

## Prerequisites

To follow along with this tutorial, you'll need:

- A [TalkJS account](https://talkjs.com/dashboard/login)
- A [Firebase account](https://console.firebase.google.com/) with a Firebase project set up. If you don't yet have a Firebase project, follow [their guides](https://firebase.google.com/docs/guides?_gl=1*qyyupu*_up*MQ..*_ga*MTQxMDg5NTA2Ni4xNzM4MjQ2OTI5*_ga_CW55HF8NVT*MTczODI0NjkyOS4xLjAuMTczODI0NjkyOS4wLjAuMA..) to get started.
- Firebase Cloud Messaging configured for your TalkJS account. For instructions on how to set this up in the TalkJS dashboard, see [Configure Firebase Cloud Messaging](https://talkjs.com/docs/Features/Notifications/Mobile_Push_Notifications/Configure_FCM/).
- The [Firebase CLI](https://firebase.google.com/docs/cli#install_the_firebase_cli)

## Set up Firebase

> In this section, you'll initialize Firebase Hosting and configure the necessary settings.

Reviewer 2: ```suggestion
In this section, you'll initialize Firebase Hosting and configure the necessary settings using the Firebase CLI.
```
We should probably add a sentence directing people to the page for installing the Firebase CLI: https://firebase.google.com/docs/cli#install_the_firebase_cli

> ### Step 1: Initialize Firebase

Reviewer 2: ```suggestion
### Step 1: Initialize Firebase Hosting
```

> Log in to Firebase if you haven't already:

Reviewer 2: ```suggestion
Log in to Firebase CLI if you haven't already:
```

Author: Hm, I would call this "Log in to Firebase with the Firebase CLI" rather than just "Log in to Firebase" (e.g. see https://firebase.google.com/docs/cli#:~:text=Log%20into%20Firebase%20using%20your%20Google%20account%20by%20running%20the%20following%20command%3A) - I've expanded it to that

```sh
firebase login
```

Next, create a new directory for your code for this project and run the following command inside it:

```sh
firebase init hosting
```

Answer the questions when prompted:

- Select to associate your new project directory with an existing Firebase project or create a new one. If you create a new one, or you haven't yet set up Firebase Cloud Messaging in the TalkJS dashboard, see [Configure Firebase Cloud Messaging](https://talkjs.com/docs/Features/Notifications/Mobile_Push_Notifications/Configure_FCM/) in our docs.
- Select to use the default (`public`) as your public directory

> - Select Yes for "Configure as a single-page app?"

Reviewer 2: They don't have to create it as a single-page app. Either option is fine. It won't affect notifications.

Author: Thanks, updated to make that clear

          appId: "<APP_ID>", // update this with your TalkJS app ID
          me: me,
        });

        var conversation = talkSession.getOrCreateConversation(
          "sample_conversation"
        );

        conversation.setParticipant(me);
        conversation.setParticipant(other);

        var chatbox = talkSession.createChatbox();

        chatbox.select(conversation);
        chatbox.mount(document.getElementById("talkjs-container"));
      });
    </script>
  </head>
  <body>
    <div id="talkjs-container" style="height: 700px"></div>
  </body>
</html>
```

Replace `<APP_ID>` with your TalkJS app ID, which you can find in the **Settings** tab of the [TalkJS dashboard](https://talkjs.com/dashboard).

Now run the following command from the root directory of the project to deploy your site:

```sh
firebase deploy --only hosting
```

This will deploy the website at a URL like `https://<FIREBASE_APP_NAME>.web.app`. You should see a chatbox with a conversation between two users:

<!-- TODO: add 2-chatbox.jpg -->
<figure class="kg-image-card">
  <img class="kg-image" src="<URL>" alt="<ALT>"/>
</figure>

### Step 2: Add the Firebase event listener

> Next, we'll load and initialize Firebase. In the `head` tag of `index.html`, add the following scripts to load the [Firebase App SDK](https://firebase.google.com/docs/reference/js/app.md?_gl=1*rwe4iq*_up*MQ..*_ga*MTEzODkxNDU1LjE3MzgzMzA0NjY.*_ga_CW55HF8NVT*MTczODMzMDQ2NS4xLjAuMTczODMzMDQ2NS4wLjAuMA..#app_package) and the [Firebase Cloud Messaging SDK](https://firebase.google.com/docs/reference/js/messaging_?_gl=1*u25wvd*_up*MQ..*_ga*MTEzODkxNDU1LjE3MzgzMzA0NjY.*_ga_CW55HF8NVT*MTczODMzMDQ2NS4xLjAuMTczODMzMDQ2NS4wLjAuMA..):

Reviewer 2: So in the example we are using the [namespaced](https://firebase.google.com/docs/web/learn-more#modular-version) of the Firebase JS SDK instead of the modular one. The difference is that the modular one would require the use of a bundler like Webpack, rollup and the likes. 

So I've edited the link to reference the namespaced webpage. Also, I've removed the random characters that Google adds to the url. Those are just for their tracking purposes.

```suggestion
Next, we'll load and initialize Firebase. In the `head` tag of `index.html`, add the following scripts to load the [Firebase App SDK](https://firebase.google.com/docs/reference/js/v8/firebase.app.App) and the [Firebase Cloud Messaging SDK](https://firebase.google.com/docs/reference/js/v8/firebase.messaging):
```

Author: Ah, got it, thanks!

        "messaging",
        "storage",
        "analytics",
        "remoteConfig",
        "performance",
      ].filter((feature) => typeof app[feature] === "function");
      loadEl.textContent = `Firebase SDK loaded with ${features.join(", ")}`;
    } catch (e) {
      console.error(e);
      loadEl.textContent = "Error loading the Firebase SDK, check the console.";
    }

    const messaging = firebase.messaging(firebase.app());

    document.getElementById("requestToken").addEventListener("click", () => {
      messaging
        .getToken({
          vapidKey: "<FIREBASE_PUBLIC_KEY>", // update this with your Firebase project public key
        })
        .then((token) => {
          window.talkSession.setPushRegistration({
            provider: "fcm",
            pushRegistrationId: token,
          });

          document.getElementById("token").textContent = token;
        });
    });
  });
</script>
```

Replace the placeholder object in `firebase.initializeApp` with the `firebaseConfig` object that you saved in Step 2 of the previous section.

Also, replace `<FIREBASE_PUBLIC_KEY>` with the public key that you saved in Step 3 of the previous section.

> When you press the **Request token** button, the event listener will get your token with Firebase's [`messaging.getToken` method](https://firebase.google.com/docs/reference/js/messaging_.md?_gl=1*13rj7jn*_up*MQ..*_ga*MjA5OTY5NTk5OS4xNzM4MjM4MTc5*_ga_CW55HF8NVT*MTczODIzODE3OC4xLjAuMTczODIzODE3OC4wLjAuMA..#gettoken_b538f38). It will then call TalkJS's [`setPushRegistration` method](https://talkjs.com/docs/Reference/JavaScript_Chat_SDK/Session/?ref=talkjs.com#Session__setPushRegistration) with the token to register for FCM push notifications.

Reviewer 2: ```suggestion
When you press the **Request token** button, the event listener will get your token with Firebase's [`messaging.getToken` method](https://firebase.google.com/docs/reference/js/v8/firebase.messaging.Messaging#gettoken). It will then call TalkJS's [`setPushRegistration` method](https://talkjs.com/docs/Reference/JavaScript_Chat_SDK/Session#Session__setPushRegistration) with the token to register for FCM push notifications.
```

  storageBucket: "",
  messagingSenderId: "",
  appId: "",
});

// Retrieve an instance of Firebase Messaging so that it can handle background
// messages.
const messaging = firebase.messaging();

// Handle incoming messages. Called when:
// - a message is received while the app has focus
// - the user clicks on an app notification created by a service worker
//   `messaging.onBackgroundMessage` handler.
messaging.onMessage((payload) => {
  console.log("Message received. ", payload);
  // ...
});

messaging.onBackgroundMessage((payload) => {
  console.log(
    "[firebase-messaging-sw.js] Received background message ",
    payload
  );

  const data = JSON.parse(payload.data["talkjs"]);
  console.log("TalkJS payload: ", data);

  // Customize notification here
  const notificationTitle = `Message from ${data.sender.name}`;
  const notificationOptions = {
    body: data.message.text,
    icon: "/firebase-logo.png",
  };

  self.registration.showNotification(notificationTitle, notificationOptions);
});
```

As with `index.html`, replace the placeholder object in `firebase.initializeApp` with the `firebaseConfig` object that you saved in Step 2 of the previous section.

> This service worker handles incoming messages and adds logging and custom notification text.

Reviewer 2: ```suggestion
This service worker handles incoming messages and only shows notifications when the webpage is in the background.
```

Should the user want, they can show notifications while the webpage is in the foreground/active. They just need to copy the code inside `onBackgroundMessage` to `onMessage`

Author: Thanks, edited to explain this

    color: rgba(0, 0, 0, 0.87);
    font-family: Roboto, Helvetica, Arial, sans-serif;
    margin: 0;
    padding: 0;
  }
  .load {
    color: #525252;
    text-align: center;
    font-size: 13px;
  }
  #requestToken {
    margin: 0px 0px 12px;
    padding: 8px 16px;
    border-radius: 4px;
    background: #007df9;
    color: white;
    border: none;
    cursor: pointer;
  }
  @media (max-width: 600px) {
    body {
      margin-top: 0;
      background: white;
      box-shadow: none;
    }
  }
</style>
```

This styles the **Request token** button and the loading

## Test your app

Let's verify that your push notification setup is working correctly by redeploying the application and sending a message.

### Step 1: Deploy your application to Firebase hosting

Redeploy your site to Firebase Hosting:

```sh
firebase deploy --only hosting
```

This will deploy your website to a URL like `https://<APP_NAME>.web.app`.

You should now see a TalkJS chatbox and a **Request token** button:

<!-- TODO: add 3-chatbox-with-token-button.jpg -->

<figure class="kg-image-card">
  <img class="kg-image" src="<URL>" alt="<ALT>"/>
</figure>

### Step 2: Send a message from another user

> TalkJS only sends notifications where you are currently not viewing a conversation, to save users from getting unnecessary notifications. For more information on exactly when TalkJS sends notifications, see [our notifications docs](https://talkjs.com/docs/Features/Notifications/#when-are-notifications-sent). So to test notifications, you will need to switch to another tab and then send a message from a different user:

Reviewer 2: This is not true. We technically send push notifications always. What determines if a notification will be shown to the user is how the user's app is configured. In this case we have setup the service worker to only show notifications when the app/webpage is in the background. i.e. the user is not looking at the webpage.

I think we need to update that docs page when it comes to push notifications

Author: Thanks, updated this and I'll look into updating the docs too

- On the device where you want to receive push notifications, go to `https://<APP_NAME>.web.app` and click the **Request token** button. Select to allow push notifications if prompted.

> - Switch to a different website in another different tab

Reviewer 2: They can even switch to a different app.

Author: Added this in
